ring/ring_test.go

```diff
@@ -2154,3 +2156,91 @@ func userToken(user, zone string, skip int) uint32 {
 	}
 	return r.Uint32()
 }
+
+func TestUpdateMetrics(t *testing.T) {
+	cfg := Config{
+		KVStore:              kv.Config{},
+		HeartbeatTimeout:     0, // get healthy stats
+		ReplicationFactor:    3,
+		ZoneAwarenessEnabled: true,
+	}
+
+	registry := prometheus.NewRegistry()
+
+	// create the ring to set up metrics, but do not start
+	ring, err := NewWithStoreClientAndStrategy(cfg, testRingName, testRingKey, nil, NewDefaultReplicationStrategy(), registry, log.NewNopLogger())
+	require.NoError(t, err)
+
+	ringDesc := Desc{
+		Ingesters: map[string]InstanceDesc{
+			"A": {Addr: "127.0.0.1", Timestamp: 2, Tokens: []uint32{math.MaxUint32 / 4, (math.MaxUint32 / 4) * 3}},
+			"B": {Addr: "127.0.0.2", Timestamp: 1, Tokens: []uint32{(math.MaxUint32 / 4) * 2, math.MaxUint32}},
+		},
+	}
+	ring.updateRingState(&ringDesc)
+	ring.updateRingMetrics()
+
+	assert.InDelta(t, 4.0, testutil.ToFloat64(ring.totalTokensGauge), 0.0001)
+
+	assert.InDelta(t, 0.5, testutil.ToFloat64(ring.memberOwnershipGaugeVec.WithLabelValues("A")), 0.0001)
+	assert.InDelta(t, 0.5, testutil.ToFloat64(ring.memberOwnershipGaugeVec.WithLabelValues("B")), 0.0001)
+	assert.InDelta(t, 2.0, testutil.ToFloat64(ring.numTokensGaugeVec.WithLabelValues("A")), 0.0001)
+	assert.InDelta(t, 2.0, testutil.ToFloat64(ring.numTokensGaugeVec.WithLabelValues("B")), 0.0001)
+	for _, s := range []string{unhealthy, ACTIVE.String(), LEAVING.String(), PENDING.String(), JOINING.String()} {
+		if s == ACTIVE.String() {
+			assert.InDelta(t, 2.0, testutil.ToFloat64(ring.numMembersGaugeVec.WithLabelValues(s)), 0.0001)
+			assert.InDelta(t, 1.0, testutil.ToFloat64(ring.oldestTimestampGaugeVec.WithLabelValues(s)), 0.0001)
+		} else {
+			assert.InDelta(t, 0.0, testutil.ToFloat64(ring.numMembersGaugeVec.WithLabelValues(s)), 0.0001)
+			assert.InDelta(t, 0.0, testutil.ToFloat64(ring.oldestTimestampGaugeVec.WithLabelValues(s)), 0.0001)
+		}
+	}
+}
+
+func TestUpdateMetricsWithRemoval(t *testing.T) {
+	cfg := Config{
+		KVStore:              kv.Config{},
+		HeartbeatTimeout:     0, // get healthy stats
+		ReplicationFactor:    3,
+		ZoneAwarenessEnabled: true,
+	}
+
+	registry := prometheus.NewRegistry()
+
+	// create the ring to set up metrics, but do not start
+	ring, err := NewWithStoreClientAndStrategy(cfg, testRingName, testRingKey, nil, NewDefaultReplicationStrategy(), registry, log.NewNopLogger())
+	require.NoError(t, err)
+
+	ringDesc := Desc{
+		Ingesters: map[string]InstanceDesc{
+			"A": {Addr: "127.0.0.1", Timestamp: 2, Tokens: []uint32{math.MaxUint32 / 4, (math.MaxUint32 / 4) * 3}},
+			"B": {Addr: "127.0.0.2", Timestamp: 1, Tokens: []uint32{(math.MaxUint32 / 4) * 2, math.MaxUint32}},
+		},
+	}
+	ring.updateRingState(&ringDesc)
+	ring.updateRingMetrics()
+
```

Comment:
How about I add those as "required" instead of "assert" ? I had multiple people in the past say that we should not do assertions on the preconditions or if we do, save the values and execute the asserts at the end (given-when-then and 4 phase test styles). And I tend to agree

Reply:
👍  On Mauro's feedback and 👍  on using `require` for pre-condition checks.

Reply:
done

Reply:
done